File: JAVA/17_Number&CombinationTheory/BOJ_05_2981_검문.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.ArrayList;
import java.util.Collections;

public class Main {

    public static int gcd(int a, int b){
        while (b != 0){
            int r = a % b;
            a = b;
            b = r;
        }
        return a;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(br.readLine());
        }
        Arrays.sort(arr);
        int d = arr[1] - arr[0];
        for (int i = 2; i < n; i++) {
            d = gcd(d, arr[i] - arr[i - 1]);
        }
        ArrayList<Integer> list = new ArrayList<>();
        for (int i = 2; i <= Math.sqrt(d); i++) {
            if (Math.pow(i, 2) == d) {
                list.add(i);
            } else if (d % i == 0) {
                list.add(i);
                list.add(d / i);
            }
        }
        list.add(d);
        Collections.sort(list);
        StringBuilder sb = new StringBuilder();
        for (int e : list) {
            sb.append(e).append(" ");
        }
        System.out.println(sb);
    }
}

File: C/17_Number&CombinationTheory/BOJ_05_2981_검문.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
int gcd(int a, int b) {
	while (b != 0) {
		int r = a % b;
		a = b;
		b = r;
	}
	return a;
}
int compare(const void* a, const void* b) {
	int o1 = *(int*)a;
	int o2 = *(int*)b;
	return o1 - o2;
}
main() {
	int n;
	scanf("%d", &n);
	int* arr = (int*)malloc(sizeof(int) * n);
	for (int i = 0; i < n; i++) {
		scanf("%d", &arr[i]);
	}
	qsort(arr, n, sizeof(int), compare);
	int d = arr[1] - arr[0];
	for (int i = 2; i < n; i++) {
		d = gcd(d, arr[i] - arr[i - 1]);
	}
	int* list = (int*)malloc(sizeof(int) * (2 * sqrt(d) + 1));
	int cnt = 0;
	for (int i = 2; i <= sqrt(d); i++) {
		if (pow(i, 2) == d) {
			list[cnt++] = i;
		}
		else if (d % i == 0) {
			list[cnt++] = i;
			list[cnt++] = d / i;
		}
	}
	list[cnt++] = d;
	qsort(list, cnt, sizeof(int), compare);
	for (int i = 0; i < cnt; i++) {
		printf("%d ", list[i]);
	}
}